File: Day2/Day2.py
input = open("Day2/input_day2.txt").readlines()
# input = ["forward 5",
# "down 5",
# "forward 8",
# "up 3",
# "down 8",
# "forward 2"]

horizontal = 0
depth = 0
aim=0

for i in input:
    command, magnitude_string = i.split()
    magnitude = int(magnitude_string)
    if command == "forward":
        horizontal = horizontal+magnitude
        depth = depth + (magnitude*aim)
    if command == "down":
        aim = aim + magnitude
    if command == "up":
        aim = aim - magnitude
        
print(horizontal)
print(depth)
print(horizontal*depth)

File: Day2/day2.hs
import System.IO
import Data.Function

readIn :: IO [String]
readIn = do
    content <- readFile "input_day2.txt"
    return (lines content)

data Direction = Forward | Up | Down 
    deriving Show

parseDirection :: String -> Maybe Direction
parseDirection "forward" = Just Forward
parseDirection "up" = Just Up
parseDirection "down" = Just Down
parseDirection _ = Nothing

data Command = Command
    { direction :: Direction
    , distance :: Int
    } deriving Show

parseCommand :: [String] -> Maybe Command
parseCommand [dir, dist] = do 
    direction <- parseDirection dir
    return (Command (direction) (read dist :: Int))

move :: (Int, Int) -> Maybe Command -> (Int, Int)
move prev Nothing = prev
move (x, y) (Just (Command Forward n)) = (x + n, y)
move (x, y) (Just (Command Up n)) = (x, y - n)
move (x, y) (Just (Command Down n)) = (x, y + n)

tupProd (a, b) = a * b

solve :: [String] -> IO ()
solve = putStrLn . show . tupProd . (foldl move (0, 0)) . (map parseCommand) . (map words)

main = readIn >>= solve
